lib/ansible/module_utils/basic.py

```diff
@@ -2685,14 +2688,17 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
         prev_dir = os.getcwd()
 
         # make sure we're in the right working directory
-        if cwd and os.path.isdir(cwd):
-            cwd = to_bytes(os.path.abspath(os.path.expanduser(cwd)), errors='surrogate_or_strict')
-            kwargs['cwd'] = cwd
-            try:
-                os.chdir(cwd)
-            except (OSError, IOError) as e:
-                self.fail_json(rc=e.errno, msg="Could not open %s, %s" % (cwd, to_native(e)),
-                               exception=traceback.format_exc())
+        if cwd:
+            if os.path.isdir(cwd):
```

Comment:
this changes the previous behaviour, do not bypass attempting execution when cwd is not a directory

Reply:
... but only when ignore_invalid_cwd is False, which is the entire point of that argument?

I admit the diff is hard to read for me, I'd go with

```python
if cwd and os.path.isdir(cwd):
    ... same code as before ...
elif cwd and not ignore_invalid_cwd:
   self.fail_json(msg="Provided cwd is not a valid directory: %s" % cwd)
```

to make the diff smaller.

Reply:
@bcoca I beg to differ. Existing code will work with this PR without any change. There was a specific test for that behaviour (bypass execution when cwd is not a directory) and it still passes. As @mgedmin pointed out, things will only change when the (new arg) ignore_invalid_cwd is False.

@mgedmin That would make the *diff* easier to read, but I believe that in the long run the PR code makes sense:
```
if cwd:         # otherwise we shouldn't bother at all with this
  if isdir:      # then do the same thing as before
  elif not ignore_invalid_cwd     # this could read "elif not isdir and not ignore_invalid_cwd", but since it is already in the if isdir stmt...
```

Reply:
keep as is, i did misread the code, but it is fine 'stand alone' it was just the diff itself that was confusing.